File: app/src/sync/fusion.ts
// Fusión por campo (sección 5.4, ADR 0002). Debe coincidir exactamente
// con server/geo/sync/fusion.py — cualquier divergencia es un bug. No
// vive en src/dominio porque conceptualmente es parte del protocolo de
// sincronización, no un motor de planificación puro, pero es igual de
// pura: sin red, sin IO, sin Date.

export const CAMPOS_CRITICOS = new Set([
  'estado',
  'vencimiento',
  'prioridad',
  'fecha_revision',
  'duracion_estimada_min',
])

export interface MetaCampo {
  updated_at: string
  device_id: string
}

export interface ConflictoCampo {
  campo: string
  valor_ganador: unknown
  device_ganador: string
  ts_ganador: string
  valor_perdedor: unknown
  device_perdedor: string
  ts_perdedor: string
}

export interface ResultadoFusionCampo {
  valor: unknown
  meta: MetaCampo
  conflicto: ConflictoCampo | null
}

function gana(candidata: MetaCampo, actual: MetaCampo): boolean {
  if (candidata.updated_at !== actual.updated_at) {
    return candidata.updated_at > actual.updated_at
  }
  return candidata.device_id > actual.device_id
}

/** Fusiona un único campo que ya tenía valor previo (con su field_meta).
 * Un campo sin valor previo no pasa por aquí: se acepta directo. */
export function fusionarCampo(
  nombreCampo: string,
  valorActual: unknown,
  metaActual: MetaCampo,
  valorEntrante: unknown,
  metaEntrante: MetaCampo,
): ResultadoFusionCampo {
  let valorGanador: unknown
  let metaGanadora: MetaCampo
  let valorPerdedor: unknown
  let metaPerdedora: MetaCampo

  if (gana(metaEntrante, metaActual)) {
    valorGanador = valorEntrante
    metaGanadora = metaEntrante
    valorPerdedor = valorActual
    metaPerdedora = metaActual
  } else {
    valorGanador = valorActual
    metaGanadora = metaActual
    valorPerdedor = valorEntrante
    metaPerdedora = metaEntrante
  }

  let conflicto: ConflictoCampo | null = null
  if (CAMPOS_CRITICOS.has(nombreCampo) && valorPerdedor !== valorGanador) {
    conflicto = {
      campo: nombreCampo,
      valor_ganador: valorGanador,
      device_ganador: metaGanadora.device_id,
      ts_ganador: metaGanadora.updated_at,
      valor_perdedor: valorPerdedor,
      device_perdedor: metaPerdedora.device_id,
      ts_perdedor: metaPerdedora.updated_at,
    }
  }

  return { valor: valorGanador, meta: metaGanadora, conflicto }
}

File: server/geo/sync/fusion.py
"""Fusión por campo (sección 5.4, ADR 0002). Puro: sin SQLAlchemy, sin
IO. Debe coincidir exactamente con app/src/sync/fusion.ts — cualquier
divergencia entre ambos es un bug, no una libertad de implementación."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

CAMPOS_CRITICOS = frozenset(
    {"estado", "vencimiento", "prioridad", "fecha_revision", "duracion_estimada_min"}
)


@dataclass(frozen=True)
class MetaCampo:
    updated_at: str
    device_id: str


@dataclass(frozen=True)
class Conflicto:
    campo: str
    valor_ganador: Any
    device_ganador: str
    ts_ganador: str
    valor_perdedor: Any
    device_perdedor: str
    ts_perdedor: str


@dataclass(frozen=True)
class ResultadoFusionCampo:
    valor: Any
    meta: MetaCampo
    conflicto: Conflicto | None


def _gana(candidata: MetaCampo, actual: MetaCampo) -> bool:
    """True si `candidata` debe reemplazar a `actual`. Determinista y
    simétrica: mismo resultado sin importar qué dispositivo evalúe."""
    if candidata.updated_at != actual.updated_at:
        return candidata.updated_at > actual.updated_at
    return candidata.device_id > actual.device_id


def fusionar_campo(
    nombre_campo: str,
    valor_actual: Any,
    meta_actual: MetaCampo,
    valor_entrante: Any,
    meta_entrante: MetaCampo,
) -> ResultadoFusionCampo:
    """Fusiona un único campo que ya tenía un valor previo (con su
    field_meta). Para un campo sin valor previo no se llama esta función:
    el valor entrante se acepta directamente, sin comparación posible."""
    if _gana(meta_entrante, meta_actual):
        valor_ganador, meta_ganadora = valor_entrante, meta_entrante
        valor_perdedor, meta_perdedora = valor_actual, meta_actual
    else:
        valor_ganador, meta_ganadora = valor_actual, meta_actual
        valor_perdedor, meta_perdedora = valor_entrante, meta_entrante

    conflicto = None
    if nombre_campo in CAMPOS_CRITICOS and valor_perdedor != valor_ganador:
        conflicto = Conflicto(
            campo=nombre_campo,
            valor_ganador=valor_ganador,
            device_ganador=meta_ganadora.device_id,
            ts_ganador=meta_ganadora.updated_at,
            valor_perdedor=valor_perdedor,
            device_perdedor=meta_perdedora.device_id,
            ts_perdedor=meta_perdedora.updated_at,
        )

    return ResultadoFusionCampo(valor=valor_ganador, meta=meta_ganadora, conflicto=conflicto)
